Fix passing of linker with target-applies-to-host and an implicit target

gmorenz opened the pull request:
### What does this PR try to resolve?

Fixes the linker half of #14195.

The second commit ("Fix `linker` value with `targeta-applies-to-host` and implicit target by storing it in `Unit`") does just that following the same structure as #13900, though the current place where `linker` is stored is a bit farther away in the code base the logical change and justification for it is the same as in #13900. As part of this we end up passing `target_config` into `TargetInfo::new` to get convenient access to the `linker` argument, there is an alternative here that I felt was worse to simply fetch the `linker` field from gctx, which is more like how `rustflags` currently works. This ends up leaving a bit of a footgun though, where that `target_config` doesn't necessarily have the right rustflags values despite being passed in.

The third commit ("Read `rustflags` from `TargetConfig` in `TargetInfo::new`") changes `rustflags` to be read from `TargetConfig` like `linker` works above. Apart from being arguably cleaner code, this removes the footgun I introduced in the second commit, and removes an existing footgun where the rustflags in `RustcTargetData::host_config` aren't necessarily correct (though it continues to be the case that `TargetConfig`s don't necessarily have the right values if there is another source of values). It was also simply made easier to do now since the second commit already conceded to making the change that `TargetInfo::new` requires a `TargetConfig`. That said, it is largely independent from this PR and could be removed if it is a sticking point.

The first commit just adds a test.

### How should we test and review this PR?

For testing, the second commit comes with it's own test (from the first commit), the third commit relies on existing tests to catch any mistakes.

### Additional information

The other half of fixing this issue is #14205, the two PRs will have a merge conflict as a result of field ordering in `Unit` but are otherwise independent.

@rustbot label: Z-target-applies-to-host

r? @weihanglo if you're willing to look at this

gmorenz commented on src/cargo/core/compiler/build_context/target_info.rs:
```diff
@@ -25,7 +25,8 @@ use std::path::{Path, PathBuf};
 
-/// Information about the platform target gleaned from querying rustc.
+/// Information about the platform target gleaned from querying rustc and from
+/// merging configs.
```
This comment didn't really fit with this struct containing `rustflags` (which predates me working on target-applies-to-host). Putting `linker` in it makes the mismatch slightly worse, so I thought I'd update the comment.

I'm not entirely thrilled with the phrasing. If someone has a suggestion for better phrasing I'd be happy to take it.

weihanglo commented:
Going to propose this change to the Cargo team on [Zulip t-cargo](https://rust-lang.zulipchat.com/#narrow/stream/246057-t-cargo) soon. Probably tomorrow.

I don't think we need to go through an FCP process for updating an unstable feature. Still no harm to get a soft consensus from the team.

gmorenz commented:
Sounds good, I'll keep a tab open to that in case there's anything I can contribute.

gmorenz commented on tests/testsuite/bad_config.rs:
```diff
@@ -21,7 +21,8 @@ fn bad1() {
         .with_stderr_data(str![[r#"
-[ERROR] expected table for configuration key `target.nonexistent-target`, but found string in [ROOT]/foo/.cargo/config.toml
+[ERROR] invalid configuration for key `target.nonexistent-target`
+expected a table, but found a string for `target.nonexistent-target` in [ROOT]/foo/.cargo/config.toml
```
Why did this test change?

Because we parse this configuration in two places, in two slightly different ways, and we return slightly different error messages depending on what parsing path we went down. Moving `target_linker` so that it's called much earlier in the process switched which code path tries to parse this first, and thus which error we get.

I don't think the errors are meaningfully different, so I'm proposing just adopting the error that happens to be returned first as correct. If anyone disagrees, I think the best alternative would be to simply make the two error paths return the same string. That's slightly awkward because the error path we are now hitting is generic over types so we'd have to either special case "table" (which wouldn't be hard) or decide we're happy with the same syntax for all types.

If there's any interest, the function call paths to these errors was:

`RustcTargetData::new -> TargetInfo::merge_compile_kind -> gctx.target_cfg_triple -> target::load_target_triple -> target::load_config_table -> gctx.get::<OptValue<PathAndArgs>>("{prefix}.runner") -> T::deserialize -> Deserializer::deserialize_option -> gctx.has_key -> self.get_cv -> self.get_cv_helper -> bail!`

And is now

`RustcTargetData::new -> TargetInfo::new -> target_linker -> gctx.target_cfgs -> target::load_target_cfgs -> gctx.get::<BTreeMap<String, TargetCfgConfig>>("target") -> T::deserialize -> ConfigMapAccess::new_map -> gctx.get_table -> gctx.expected`

bors commented:
:umbrella: The latest upstream changes (presumably #14243) made this pull request unmergeable. Please [resolve the merge conflicts](https://rustc-dev-guide.rust-lang.org/git.html#rebasing-and-conflicts).

gmorenz commented:
Huh, why isn't CI running?

Rebased on master (at least as of a few hours ago) and changed from using `Arc` to `Rc` as suggested by weihanglo in #14205.

weihanglo commented:
Perhaps <https://www.githubstatus.com/incidents/m70hk23gx3nx>

gmorenz commented:
That's a better guess than mine that github was unhappy about the fact that there was a few hours between me making the changes and pushing them.

Rebased on master again (no longer 4 hours in the past).

gmorenz commented on src/cargo/core/compiler/build_context/target_info.rs:
```diff
@@ -54,6 +56,8 @@ pub struct TargetInfo {
     pub rustflags: Arc<[String]>,
     /// Extra flags to pass to `rustdoc`, see [`extra_args`].
     pub rustdocflags: Arc<[String]>,
```
Yes, will do in a separate PR.

gmorenz commented on src/cargo/core/compiler/build_context/target_info.rs:
```diff
@@ -940,9 +928,14 @@ impl<'gctx> RustcTargetData<'gctx> {
         let host_config = if target_applies_to_host {
-            gctx.target_cfg_triple(&rustc.host)?
+            let mut config = gctx.target_cfg_triple(&rustc.host)?;
+            if requested_kinds != [CompileKind::Host] {
+                // When an explicit target flag is passed rustflags are ignored for host artifacts.
+                config.rustflags = None;
+                config.rustdocflags = None;
+            }
```
Yes, exactly.

And I 100% agree with bad code smell, but this less bad to me than the pre-existing reality that host_config is just storing incorrect rustflags in this case. It's one of the things that took me awhile to understand the first time I started working on this code (and why I added that comment on line 943 that I now get to delete).

A more ambitious refactoring option might be to try and push all this logic into something like `gctx.host_cfg_triple`. I remember trying that when I wrote this and rejecting it, but I don't remember why I rejected it and I'd be happy to give it another shot if you think that it might be cleaner.

gmorenz commented on src/cargo/core/compiler/build_context/target_info.rs:
```diff
@@ -826,6 +836,42 @@ fn rustflags_from_target(
 }
 
+/// Gets the user-specified linker for a particular host or target from the configuration.
+fn target_linker(
```
Good catch! I entirely missed that this was a problem.

I don't see a good solution to this, I'll come back with a fresh set of eyes tomorrow and see if that changes.

Right now my take is it's either this (bad), don't fix this issue at all (also bad), or doing something involving not so clean code to defer actually returning the error (also bad). But I think there's a small chance I can find a better way.

weihanglo replied:
Honestly, the current approach is not too bad for normal users.

However, I know some enterprise users have weird multi-level configuration setup. And may move projects among different directories. This is also bad because you cannot opt out directory-probing for configuration files (which Cargo may provide an escape patch?).

bors commented:
:umbrella: The latest upstream changes (presumably #14273) made this pull request unmergeable. Please [resolve the merge conflicts](https://rustc-dev-guide.rust-lang.org/git.html#rebasing-and-conflicts).
